### ISSUE TITLE
computing probability involving a weibull distribution raises ValueError

### ISSUE THREAD

**Issue by l-johnston** [sympy/sympy#21236]
OS: Linux Ubuntu 20.04
Python: 3.9.3
SymPy: 1.7.1

```python
>>> import sympy
>>> sympy.__version__
'1.7.1'
>>> from sympy.stats import Weibull, P
>>> X = Weibull("x", 8000, 1.5)
>>> P(X <= 4000)
...
ValueError: x**w where w is irrational is not defined for negative x
```
The problem appears to be with the shape factor being a floating point value causing SymPy to interpret the value 1.5 as an irrational number. Forcing the shape factor to be Rational gives the expected result.

```python
>>> from sympy import Rational, N
>>> X = Weibull("x", 8000, Rational(3, 2))
>>> P(X <= 4000)
1 - exp(-sqrt(2)/4)
>>> N(_)
0.297811498673440
```
Working directly with the cdf also gives the expected result.

```python
>>> from sympy.stats import cdf
>>> X = Weibull("x", 8000, 1.5)
>>> cdf(X)(4000)
0.297811498673440
```

This problem does not appear in SymPy version 1.5.1.

```python
>>> import sympy
>>> sympy.__version__
'1.5.1'
>>> from sympy.stats import Weibull, P
>>> X = Weibull("x", 8000, 1.5)
>>> P(X <= 4000)
0.297811498673440
```

**Comment by golmdav** [comment]
I'd like to work on this issue.

**Comment by oscarbenjamin** [comment]
The particular issue described here could be fixed by this diff:
```diff
diff --git a/sympy/solvers/solveset.py b/sympy/solvers/solveset.py
index 17da5ae025..786041be5e 100644
--- a/sympy/solvers/solveset.py
+++ b/sympy/solvers/solveset.py
@@ -258,7 +258,7 @@ def _invert_real(f, g_ys, symbol):
                 else:
                     return _invert_real(base, res, symbol)
             else:
-                if not base.is_positive:
+                if base.is_positive is False:
                     raise ValueError("x**w where w is irrational is not "
                                      "defined for negative x")
                 return _invert_real(base, res, symbol)
```
The problem is that in this case `base` is a `Symbol` with `real=True` and it is not known to be positive:
```python
In [1]: base = Symbol('base', real=True)

In [2]: print(base.is_positive)
None
```
The `is_positive` property is part of the core assumptions and uses three-valued logic: True, False or None with None meaning unknown. It is not usually valid to use the plain Python `not` operator with a three-valued bool because `not None` gives `True`. In this case that means that an error is raised if the `base` is not known to be positive which is probably not correct.

A simpler way to see the error is:
```python
In [13]: solveset(x**1.5, x, Reals)
---------------------------------------------------------------------------
ValueError: x**w where w is irrational is not defined for negative x
```

**Comment by l-johnston** [comment]
@oscarbenjamin , as usual things are more complicated. `solveset` strips assumptions so that `base.is_positive` is always `None` no matter what the end user provides to `solveset`. While the change you propose works for the original problem, it breaks the existing test suite and does not allow for the case where `base` is explicitly non-positive.

**Comment by oscarbenjamin** [comment]
So we're trying to rearrange
```
f(x)**w  in  B
```
 into
```
f(x) in {real_root(b, w) where b in B}
```
 where `x`, `f(x)`, `w`, are real and `B` is a subset of the Reals and `w` is irrational.

If `f(x)` is positive (for every real value of x) then that's simple. If `f(x)`.

Now consider if `f(x)` is negative for some x. We can write `f(x)` as `|f(x)| * exp(I*pi)` and hence
```
f(x)**w = |f(x)|**w * exp(I*pi)**w
```
Here `exp(I*pi)**w` can never be real if `w` is irrational because it has argument `pi*w + 2*n*pi`. That means that there can be no solutions where `f(x) < 0`.

So maybe the solution is something like
```
f(x) in {real_root(b, w), b in B} & Interval(0, oo)
```
In other words rather than raising an error we can adjoin this condition on `f(x)` to the rhs set `res`.

**Comment by oscarbenjamin** [comment]
I think this does it:
```diff
diff --git a/sympy/solvers/solveset.py b/sympy/solvers/solveset.py
index aecdb2c2cb..6c654cf708 100644
--- a/sympy/solvers/solveset.py
+++ b/sympy/solvers/solveset.py
@@ -257,10 +257,9 @@ def _invert_real(f, g_ys, symbol):
 
                 else:
                     return _invert_real(base, res, symbol)
-            else:
-                if not base.is_positive:
-                    raise ValueError("x**w where w is irrational is not "
-                                     "defined for negative x")
+            elif expo.is_irrational or expo.is_Float:
+                g_ys_pos = g_ys & Interval(0, oo)
+                res = imageset(Lambda(n, real_root(n, expo)), g_ys_pos)
                 return _invert_real(base, res, symbol)
 
         if not base_has_sym:
```
That gives e.g.:
```python
In [7]: solveset(x**pi-E, x, Reals)
Out[7]: 
⎧ 1⎫
⎪ ─⎪
⎨ π⎬
⎪ℯ ⎪
⎩  ⎭
```

**Comment by l-johnston** [comment]
@oscarbenjamin, well not quite. What about the case when `expo=1.0`? With your proposed solution:
```python
>>> solveset(x**1.0 + 1, x, Reals)
EmptySet
>>> solveset(x**1 + 1, x, Reals)
FiniteSet(-1)
```

**Comment by l-johnston** [comment]
@oscarbenjamin, moving the `Float` case up appears to resolve the previous issue:

```diff
diff --git a/a/sympy/solvers/solveset.py b/b/sympy/solvers/solveset.py
index 4eb82a9..40683a7 100644
--- a/a/sympy/solvers/solveset.py
+++ b/b/sympy/solvers/solveset.py
@@ -240,7 +240,7 @@ def _invert_real(f, g_ys, symbol):

         if not expo_has_sym:
             res = imageset(Lambda(n, real_root(n, expo)), g_ys)
-            if expo.is_rational:
+            if expo.is_rational or expo.is_Float:
                 numer, denom = expo.as_numer_denom()
                 if denom % 2 == 0:
                     base_positive = solveset(base >= 0, symbol, S.Reals)
@@ -257,10 +257,9 @@ def _invert_real(f, g_ys, symbol):

                 else:
                     return _invert_real(base, res, symbol)
-            else:
-                if not base.is_positive:
-                    raise ValueError("x**w where w is irrational is not "
-                                     "defined for negative x")
+            elif expo.is_irrational:
+                g_ys_pos = g_ys & Interval(0, oo)
+                res = imageset(Lambda(n, real_root(n, expo)), g_ys_pos)
                 return _invert_real(base, res, symbol)

         if not base_has_sym:
```

```python
>>> solveset(x**1.0 + 1, x, Reals)
FiniteSet(-1.0)
```

**Comment by golmdav** [comment]
@oscarbenjamin I have made your most recent suggested changes on my build and

solveset(x**1.0 + 1, x, Reals) -> FiniteSet(-1.0)
solveset(x**pi-E, x, Reals) -> FiniteSet(exp(1/pi))
solveset(x**1.5, x, Reals) -> FiniteSet(0)

all of these tests appears to return the correct answer.

Is there anything I can do to make this change happen? This is my first issue and I'd like to make a contribution somehow. I could potentially add these tests and a few more into the regression test suite.

**Comment by oscarbenjamin** [comment]
You can send a pull request with these changes. It would be good to have some tests for cases with either a non-trivial base or a non-trivial rhs e.g. if the rhs is always/sometimes negative or something like that. I expect that there are already tests for this somewhere so these can be added to those test functions along with a comment pointing to this issue e.g.:
```
   # https://github.com/sympy/sympy/issues/21236
```

**Comment by golmdav** [comment]
Alright, was testing out the fix with some less trivial functions.
@l-johnston , I've found a test that your last fix does not pass:

solveset(sin(x)**1.0, x, Reals) -> ImageSet(Lambda(_n, 2*_n*pi), Integers)

If we change 1.0 to an int we get:

solveset(sin(x)**1, x, Reals) -> Union(ImageSet(Lambda(_n, 2*_n*pi + pi), Integers), ImageSet(Lambda(_n, 2*_n*pi), Integers))

Which is the correct answer

Not sure what is happening here though.

**Comment by oscarbenjamin** [comment]
> Not sure what is happening here though.

The problem might be somewhere else. Try to find out exactly what `invert` is doing and let's at least make sure that part is correct.

**Comment by golmdav** [comment]
yeah you're right, the sin problem is elsewhere because when I run,

`invert_real(sin(x)**1, 0, x)
invert_real(sin(x)**1.0, 0, x)`

I get 

`(x, ImageSet(Lambda(_n, _n*pi), Integers))
(x, ImageSet(Lambda(_n, _n*pi), Integers))`

which are both correct
